File: dgs/src/test/java/com/example/demo/MutationTests.java
package com.example.demo;

import com.example.demo.gql.datafetchers.PostsDataFetcher;
import com.example.demo.gql.types.CreatePostInput;
import com.example.demo.gql.types.Post;
import com.example.demo.service.AuthorService;
import com.example.demo.service.PostService;
import com.netflix.graphql.dgs.DgsQueryExecutor;
import com.netflix.graphql.dgs.test.EnableDgsTest;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@EnableDgsTest
@SpringBootTest(classes = {MutationTests.MutationTestsConfig.class})
@Slf4j
class MutationTests {

    @Configuration
    @Import(PostsDataFetcher.class)
    static class MutationTestsConfig {

    }

    @Autowired
    DgsQueryExecutor dgsQueryExecutor;

    @MockitoBean
    PostService mockedPostService;

    @MockitoBean
    AuthorService authorService;

    @Test
    void createPosts() {
        when(mockedPostService.createPost(any(CreatePostInput.class))).thenReturn(1L);
        when(mockedPostService.getPostById(anyLong())).thenReturn(
                Post.builder().id(1L)
                        .title("test title")
                        .content("test content")
                        .build()
        );

        String title = dgsQueryExecutor.executeAndExtractJsonPath(
                """
                        mutation createPost($input: CreatePostInput!){
                            createPost(createPostInput:$input){
                                id
                                title
                                content
                            }
                        }""",
                "data.createPost.title",
                Map.of("input", Map.of(
                                "title", "test title",
                                "content", "test content"
                        )
                )
        );

        assertThat(title).isEqualTo("test title");

        verify(mockedPostService, times(1)).createPost(any(CreatePostInput.class));
        verify(mockedPostService, times(1)).getPostById(anyLong());
        verifyNoMoreInteractions(mockedPostService);
    }

}

File: dgs-kotlin-co/src/test/kotlin/com/example/demo/MutationTests.kt
package com.example.demo

import com.example.demo.gql.datafetcher.PostsDataFetcher
import com.example.demo.gql.scalar.LocalDateTimeScalar
import com.example.demo.gql.scalars.UUIDScalar
import com.example.demo.gql.types.Post
import com.example.demo.service.PostService
import com.netflix.graphql.dgs.reactive.DgsReactiveQueryExecutor
import com.netflix.graphql.dgs.test.EnableDgsTest
import com.ninjasquad.springmockk.MockkBean
import graphql.schema.GraphQLScalarType
import graphql.schema.idl.RuntimeWiring
import io.kotest.matchers.shouldBe
import io.mockk.coEvery
import io.mockk.coVerify
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.reactor.awaitSingle
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.ImportAutoConfiguration
import org.springframework.boot.autoconfigure.web.reactive.WebFluxAutoConfiguration
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.graphql.execution.RuntimeWiringConfigurer
import java.util.*

@OptIn(ExperimentalCoroutinesApi::class)
@SpringBootTest(classes = [MutationTests.TestConfig::class])
@EnableDgsTest
class MutationTests {

    @Configuration
    @Import(
        value = [
            PostsDataFetcher::class,
            LocalDateTimeScalar::class,
            UUIDScalar::class
        ]
    )
    @ImportAutoConfiguration(
        value = [
            WebFluxAutoConfiguration::class
        ]
    )
    class TestConfig{

        @Bean
        fun customRuntimeWiring(): RuntimeWiringConfigurer{
            return object: RuntimeWiringConfigurer {
                override fun configure(builder: RuntimeWiring.Builder) {
                    builder.scalar(
                        GraphQLScalarType.newScalar()
                            .name("UUID")
                            .description("UUID type")
                            .coercing(UUIDScalar())
                            .build()
                    )
                }
            }
        }
    }

    @Autowired
    lateinit var dgsQueryExecutor: DgsReactiveQueryExecutor


    @MockkBean
    lateinit var postService: PostService

    @Test
    fun `create new post`() = runTest {
        val post = Post(
            id = UUID.randomUUID(),
            title = "test title",
            content = "test content"
        )
        coEvery { postService.createPost(any()) } returns post

        val query = """
            mutation createPost(${'$'}input: CreatePostInput!){
                 createPost(createPostInput:${'$'}input) 
                 {
                     id
                     title
                 } 
             }
        """.trimIndent()

        val variables = mapOf(
            "input" to mapOf(
                "title" to "test title",
                "content" to "test content"
            )
        )

        val titles = dgsQueryExecutor
            .executeAndExtractJsonPath<String>(
                query,
                "data.createPost.title",
                variables
            )
            .awaitSingle()

        titles shouldBe post.title

        // verify
        coVerify(atLeast = 1) { postService.createPost(any()) }
    }
}


